app/src/main/java/app/revanced/manager/ui/screen/PatcherScreen.kt

```diff
@@ -0,0 +1,424 @@
+                .clickable { expanded = !expanded }
+                .background(categoryColor)
+        ) {
+            Row(
+                verticalAlignment = Alignment.CenterVertically,
+                horizontalArrangement = Arrangement.spacedBy(12.dp),
+                modifier = Modifier.padding(16.dp)
+            ) {
+                StepIcon(
+                    state = when {
+                        steps.all { it.state == State.COMPLETED } -> State.COMPLETED
+                        steps.any { it.state == State.FAILED } -> State.FAILED
+                        steps.any { it.state == State.RUNNING } -> State.RUNNING
+                        else -> State.WAITING
+                    },
+                    size = 24.dp
+                )
+
+                Text(
+                    stringResource(category.displayName),
+                    style = MaterialTheme.typography.titleMedium
+                )
+
+                Spacer(modifier = Modifier.weight(1f))
+
+                Text(
+                    "${steps.count { it.state == State.COMPLETED }}/${steps.size}",
+                    style = MaterialTheme.typography.labelSmall
+                )
+
+                Arrow(modifier = Modifier.size(24.dp), expanded = expanded)
+            }
+        }
+
+        // Sub Steps
+        AnimatedVisibility(visible = expanded) {
+            Column(
+                modifier = Modifier.fillMaxWidth()
+            ) {
+                if (category != StepCategory.PATCHING) {
+                    steps.forEach { step ->
+                        val progress = step.progress?.collectAsStateWithLifecycle()
+
+                        SubStep(
+                            name = step.name,
+                            state = step.state,
+                            message = step.message,
+                            progress = progress?.value
+                        )
+                    }
+                } else {
+                    val currentPatch = steps.find { it.state == State.RUNNING }?.name
+                    val failedPatch = steps.find { it.state == State.FAILED }?.name
+
+                    SubStep(
+                        name = when {
+                            steps.any { it.state == State.RUNNING } -> "Applying $currentPatch"
+                            steps.any { it.state == State.FAILED } -> "Failed to apply $failedPatch"
+                            steps.all { it.state == State.WAITING } -> "Apply patches"
+                            else -> "Applied ${steps.size} patches"
+                        },
+                        state = when {
+                            steps.all { it.state == State.COMPLETED } -> State.COMPLETED
+                            steps.any { it.state == State.FAILED } -> State.FAILED
+                            steps.any { it.state == State.RUNNING } -> State.RUNNING
+                            else -> State.WAITING
+                        },
```

Comment:
Use `remember`. Try to determine the category state once to reduce the amount of times you are iterating over the substeps.